models/DataObject/ClassDefinition/Data/Extension/Relation.php

```diff
@@ -111,4 +120,27 @@ public function getObjectsAllowed(): bool
     {
         return false;
     }
+
+    /**
+     * @param 'asset'|'document'|'object' $type
+     */
+    private function resolveClassName(TypeLoader|AssetTypeLoader $loader, string $type, string $shortName): ?string
+    {
+        $factory = \Pimcore::getContainer()->get('pimcore.model.factory');
+        $className = match ($type) {
+            'asset' => '\\Pimcore\Model\Asset\\' . ucfirst($shortName),
+            'document' => '\\Pimcore\Model\Document\\' . ucfirst($shortName),
+            'object' => '\\Pimcore\Model\DataObject\\' . ucfirst($shortName),
+        };
+
+        try {
+            return $factory->getClassNameFor($className);
+        } catch (UnsupportedException) {
+            try {
+                return '\\' . $loader->getClassNameFor($shortName);
+            } catch (UnsupportedException) {
+                return null;
+            }
```

Comment:
Do factories really have the priority here? My assumption was that loaders are the new, preferred variant (even though I haven't quite understood the difference yet, as both seem to be based on the same implementation).

Reply:
Tbh we did not really talk about the prioritization, but we'll discuss it internally. You raised a good point though.

The prio would only be problematic if you override class and have a custom type at the same time. I don't really see a use case for that, since you can also "override" existing types with your custom type.  

For the implementation part, they are based on the implementation loaders but are quite different. 
The class_overrides work with the FQNS and you can literally override any class. 
The type loaders work with the corresponding type and ensure that you can also add new types, e.g. you want a new document type "book". 
This class will then also be available for e.g. the relations in a dataobject, it can be easier shown in the tree etc. 
You cannot do this with the class_overrides. 

We actually needed this for the Web2PrintBundle since we removed the Printcontainer and the Printpage from the core.